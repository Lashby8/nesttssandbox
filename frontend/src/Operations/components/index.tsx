export { default as OperationsTable } from "./OperationsTable";
