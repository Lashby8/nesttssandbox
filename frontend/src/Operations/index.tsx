export { default as OperationsView } from "./OperationsView";
