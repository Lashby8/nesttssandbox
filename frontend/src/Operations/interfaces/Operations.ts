export enum OperationsStatus {
  PENDING = "pending",
  DONE = "done",
  FAILED = "failed",
}

export interface Operation {
  id: string;
  name: string;
  status: OperationsStatus;
}
