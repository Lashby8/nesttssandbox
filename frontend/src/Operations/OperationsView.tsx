import React, { PureComponent } from "react";
import { Paper, Button } from "@mui/material";

import { fetchAllOperations } from "./store";
import { CreateOperationModal } from "./components/modals";
import { OperationsTable } from "./components";
import { Operation } from "./interfaces/Operations";
import { OperationsService } from "./services/OperationsServices";

export interface OperationsProps {
  operations: Operation[];
}

interface OperationsState {
  isModalOpen: boolean;
  newOperationName: string;
}

export default class OperationsView extends PureComponent<
  OperationsProps,
  OperationsState
> {
  constructor(props: OperationsProps) {
    super(props);

    this.state = {
      isModalOpen: false,
      newOperationName: "",
    };
  }

  public async componentDidMount() {
    try {
      await fetchAllOperations();
    } catch (e) {
      console.error(e);
      return;
    }
  }

  public render() {
    const { operations } = this.props;
    const { isModalOpen, newOperationName } = this.state;

    return (
      <Paper
        elevation={6}
        style={{
          height: "95vh",
          width: "95vw",
          display: "flex",
          flexDirection: "column",
          alignItems: "center",
          alignSelf: "center",
          marginLeft: "3vw",
          marginTop: "3vh",
        }}
      >
        <Button onClick={this.handleOpenCreteOperationModal}>Create</Button>
        <CreateOperationModal
          isOpened={isModalOpen}
          handleCloseAction={this.handleCloseCreteOperationModal}
          handleCreateNewOperation={async () => {
            try {
              if (newOperationName.length === 0) {
                return;
              }

              await this.handleCreateNewOperation(newOperationName);
              await fetchAllOperations();
              this.handleCloseCreteOperationModal();
              this.setState({ newOperationName: "" });
            } catch (e) {
              console.error(e);
            }
          }}
          newOperationName={newOperationName}
          handleChangeNewName={this.handleChangeNewName}
        />
        <OperationsTable operations={operations} />
      </Paper>
    );
  }

  private handleOpenCreteOperationModal = () => {
    this.setState({
      isModalOpen: true,
    });
  };

  private handleCloseCreteOperationModal = () => {
    this.setState({
      isModalOpen: false,
      newOperationName: "",
    });
  };

  handleChangeNewName = (event: React.ChangeEvent<HTMLInputElement>): void => {
    this.setState({ newOperationName: event.target.value });
  };

  private async handleCreateNewOperation(newName: string): Promise<any> {
    try {
      const operation = await new OperationsService(
        "http://localhost",
        8000
      ).createOperation(newName);
      return operation;
    } catch (e) {
      console.error(e);
      return;
    }
  }
}
