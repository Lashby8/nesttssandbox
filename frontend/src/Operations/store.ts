import { createStore, createEvent, createEffect } from "effector";
import { Operation } from "./interfaces/Operations";
import { OperationsService } from "./services/OperationsServices";

export const addOperation = createEvent<Operation>();
export const addOperations = createEvent<Operation[]>();
export const resetOperations = createEvent();
export const updateOrAddOperation = createEvent<Operation>();

export const updateOrAddOperationHandler = (
  state: Operation[],
  newOperation: Operation
): Operation[] => {
  let wasElementFound: boolean = false;

  const updatedState = state.map((oldOperation: Operation) => {
    if (oldOperation.id === newOperation.id) {
      wasElementFound = true;

      return newOperation;
    }

    return oldOperation;
  });

  return wasElementFound ? updatedState : [...state, newOperation];
};

export const fetchAllOperations = createEffect(async () => {
  const operations = await new OperationsService(
    "http://localhost",
    8000
  ).fetchAllOperations();
  return operations;
});

export const operationsStore = createStore<Operation[]>([])
  .on(addOperation, (state, operation) => [...state, operation])
  .on(addOperations, (state, operations) => [...state, ...operations])
  .on(updateOrAddOperation, updateOrAddOperationHandler)
  .on(fetchAllOperations.doneData, (_, operations) => [...operations])
  .reset(resetOperations);
