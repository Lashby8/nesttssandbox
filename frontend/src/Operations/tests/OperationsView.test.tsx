import React from "react";
import renderer from "react-test-renderer";
import OperationsView from "../OperationsView";
import { Operation, OperationsStatus } from "../interfaces/Operations";

test("OperationsView is rendered", () => {
  const operations: Operation[] = [
    {
      id: "1",
      name: "a",
      status: OperationsStatus.DONE,
    },
    {
      id: "2",
      name: "b",
      status: OperationsStatus.FAILED,
    },
  ];
  const component = renderer.create(<OperationsView operations={operations} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
