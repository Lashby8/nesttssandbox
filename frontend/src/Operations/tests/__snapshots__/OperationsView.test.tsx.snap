// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OperationsView is rendered 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation6 css-azc51k-MuiPaper-root"
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "display": "flex",
      "flexDirection": "column",
      "height": "95vh",
      "marginLeft": "3vw",
      "marginTop": "3vh",
      "width": "95vw",
    }
  }
>
  <button
    className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-1e6y48t-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Create
  </button>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
  >
    <table
      aria-label="simple table"
      className="MuiTable-root css-quj9j6-MuiTable-root"
      role={null}
    >
      <thead
        className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Name
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
            scope="col"
          >
            Status
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root css-34nofg-MuiTableRow-root"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            scope="row"
          >
            a
          </th>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            done
          </td>
        </tr>
        <tr
          className="MuiTableRow-root css-34nofg-MuiTableRow-root"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            scope="row"
          >
            b
          </th>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            failed
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
